### PR TITLE
Support push without payload

### PR DESCRIPTION
**Pull request by wibblymat**
Needed to ping client's for which we do have an endpoint but don't have encryption values.

R: @gauntface @crhym3

### CODE REVIEW

File: webpush/encrypt.go [hunk 1]
```diff
@@ -109,7 +109,10 @@ func SubscriptionFromJSON(b []byte) (*Subscription, error) {
 			Auth   string
 		}
 	}
-	json.Unmarshal(b, &sub)
+	err := json.Unmarshal(b, &sub)
+	if err != nil {
+		return nil, err
+	}
```
Comment (x1ddos) [inline webpush/encrypt.go:114]: could be written a bit shorter here:

```
if err := json.Unmarshal(b, &sub); err != nil {
  return nil, err
}
```

File: webpush/push_test.go [hunk 2]
```diff
@@ -92,3 +91,42 @@ func TestSendWebPush(t *testing.T) {
...
+		if len(body) != 0 {
+			t.Errorf("Expected body to be length 0, was %d", len(body))
+		}
+	}))
+	defer ts.Close()
+
+	serverURL, _ := url.Parse(ts.URL)
+
+	// Make a fake transport that redirects all requests to the fake server
+	transport := &FakeTransport{serverURL}
```
Comment (x1ddos) [inline webpush/push_test.go]: I don't think `FakeTransport` needs to be exported?

File: webpush/push_test.go [hunk 3]
```diff
@@ -92,3 +91,42 @@ func TestSendWebPush(t *testing.T) {
...
+	client := &http.Client{Transport: transport}
+
+	subscriptionJSON := []byte(`{
+		"endpoint": "https://example.com/"
```
Comment (x1ddos) [inline webpush/push_test.go]: Just FYI: an alternative is something like `"endpoint": ts.URL` which wouldn't require `fakeTransport` at all.

File: webpush/push_test.go [hunk 4]
```diff
@@ -92,3 +91,42 @@ func TestSendWebPush(t *testing.T) {
...
+	}`)
+
+	sub, err := SubscriptionFromJSON(subscriptionJSON)
```
Comment (x1ddos) [inline webpush/push_test.go]: You now have a dedicated test for SubscriptionFromJSON, right. So, you don't really need to verify it here. Wouldn't creating sub with `sub := &Subscription{Endpoint: ts.URL}` be easier for this test?

**Comment by wibblymat** [comment]
PTAL @crhym3

File: webpush/push.go [hunk 5]
```diff
@@ -32,24 +32,36 @@ const (
...
-	req.Header.Add("Content-Encoding", "aesgcm")
+		req, err = http.NewRequest("POST", endpoint, bytes.NewReader(payload.Ciphertext))
+		if err != nil {
+			return nil, err
+		}
+
+		req.Header.Add("Encryption", headerField("salt", payload.Salt))
+		req.Header.Add("Crypto-Key", headerField("dh", payload.ServerPublicKey))
+		req.Header.Add("Content-Encoding", "aesgcm")
+	} else {
```
Comment (x1ddos) [inline webpush/push.go]: Just a thought: wouldn't it be more readable if the `if` conditions were reversed?

``` go
req, err := http.NewRequest("POST", endpoint, nil)
if err != nil {
  return nil, err
}
if token != "" {
  req.Header.Add("Authorization", "key="+token)
}
if message == "" {
  return req, nil
}

// encrypted payload
payload, err := Encrypt(sub, message)
if err != nil {
  return nil, err
}
req.Body = ioutil.NopCloser(bytes.NewReader(payload.Ciphertext))
req.ContentLength = len(payload.Ciphertext)
req.Header.Add("Encryption", headerField("salt", payload.Salt))
req.Header.Add("Crypto-Key", headerField("dh", payload.ServerPublicKey))
req.Header.Add("Content-Encoding", "aesgcm")
return req, nil
```

File: webpush/push_test.go [hunk 6]
```diff
@@ -67,27 +56,47 @@ func TestSendWebPush(t *testing.T) {
 	}))
 	defer ts.Close()
 
-	serverURL, _ := url.Parse(ts.URL)
+	client := &http.Client{}
```
Comment (x1ddos) [inline webpush/push_test.go]: > If the client is nil then the default HTTP client will be used

you wrote it :)
`Send(nil, sub, message)`

File: webpush/push_test.go [hunk 7]
```diff
@@ -67,27 +56,47 @@ func TestSendWebPush(t *testing.T) {
...
-		"keys": {
-			"p256dh": "BCXJI0VW7evda9ldlo18MuHhgQVxWbd0dGmUfpQedaD7KDjB8sGWX5iiP7lkjxi-A02b8Fi3BMWWLoo3b4Tdl-c=",
-			"auth": "WPF9D0bTVZCV2pXSgj6Zug=="
-		}
-	}`)
+	sub := &Subscription{ts.URL, key, auth}
 	message := "I am the walrus"
 
-	sub, err := SubscriptionFromJSON(subscriptionJSON)
+	_, err = Send(client, sub, message, "")
```
Comment (x1ddos) [inline webpush/push_test.go]:
``` go
if _, err := Send(nil, sub, message, ""); err != nil {
  t.Error(err)
}
```

**Comment by wibblymat** [comment]
Done, PTAL @crhym3 :)

**Comment by x1ddos** [comment]
LGTM :star2: 
Personally, I'd squash commits before merging but up to you.

**Pull request merged**